signature/internal/base/envelope.go

```diff
@@ -0,0 +1,221 @@
+package base
+
+import (
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"github.com/notaryproject/notation-core-go/signature"
+	nx509 "github.com/notaryproject/notation-core-go/x509"
+)
+
+// Envelope represents a general envelope wrapping a raw signature and envelope
+// in specific format.
+// Envelope manipulates the common validation shared by internal envelopes.
+type Envelope struct {
+	signature.Envelope        // internal envelope in a specific format(e.g. Cose, JWS)
+	Raw                []byte // raw signature
+}
+
+// Sign generates signature in terms of given SignRequest.
+//
+// Reference: https://github.com/notaryproject/notaryproject/blob/main/signing-and-verification-workflow.md#signing-steps
+func (e *Envelope) Sign(req *signature.SignRequest) ([]byte, error) {
+	// Canonicalize request.
+	req.SigningTime = req.SigningTime.Truncate(time.Second)
+	req.Expiry = req.Expiry.Truncate(time.Second)
+	err := validateSignRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := e.Envelope.Sign(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate certificate chain
+	content, err := e.Envelope.Content()
+	if err != nil {
+		return nil, err
+	}
+	signerInfo := content.SignerInfo
```

Comment:
nit: I know this is for readability but we really don't need to have a copy of `SignerInfo`.